src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions
+{
+    internal static class CodeActionHelpers
+    {
+        public static async Task<IEnumerable<CodeAction>> GetCodeActionsAsync(
+            Document? document,
+            ICodeFixService codeFixService,
+            ICodeRefactoringService codeRefactoringService,
+            LSP.Range selection,
+            CancellationToken cancellationToken)
+        {
+            if (document == null)
+            {
+                return ImmutableArray<CodeAction>.Empty;
+            }
+
+            var (codeFixCollections, codeRefactorings) = await GetCodeFixesAndRefactoringsAsync(
+                document, codeFixService,
+                codeRefactoringService, selection,
+                cancellationToken).ConfigureAwait(false);
+
+            var codeActions = codeFixCollections.SelectMany(c => c.Fixes.Select(f => f.Action)).Concat(
+                    codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
+
+            return codeActions;
+        }
+
+        public static async Task<(ImmutableArray<CodeFixCollection>, ImmutableArray<CodeRefactoring>)> GetCodeFixesAndRefactoringsAsync(
+            Document document,
+            ICodeFixService codeFixService,
+            ICodeRefactoringService codeRefactoringService,
+            LSP.Range selection,
+            CancellationToken cancellationToken)
+        {
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
+
+            var codeFixCollections = Array.Empty<CodeFixCollection>().ToImmutableArray();
+            var codeRefactorings = Array.Empty<CodeRefactoring>().ToImmutableArray();
```

Comment:
```suggestion
            var codeRefactorings = ImmutableArray<CodeRefactoring>.Empty;
```

(applies to other locations as well).